# nf-core/hicar: Output

## Introduction

The pipeline is built using [Nextflow](https://www.nextflow.io/). See [`main README.md`](../README.md) for a condensed overview of the steps in the pipeline, and the bioinformatics tools used at each step.

See [Illumina website](https://emea.illumina.com/techniques/sequencing/dna-sequencing/chip-seq.html) for more information regarding the ChIP-seq protocol, and for an extensive list of publications.

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

<!-- TODO nf-core: Write this documentation describing your workflow's output -->

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/)
and saves processed data in the following folders:

* [FastQC](#fastqc) - Read quality control
* [MultiQC](#multiqc) - Aggregate report describing results from the whole pipeline
* [pipeline_info](#pipeline-information) - Report metrics generated during the workflow execution
* [genome](#genome) - The genome related files used in the pipeline.
* [trimmed_fq](#adapter-trimming) - The adapter trimmed fastq files.
* [bwa](#alignment) - The alignments of trimmed reads.
* [pairs](#pairs) - The interaction pairs and their quality control reports.
* [cooler](#cooler) - Cooler files for visualization
* [ATACpeak](#call-peaks-for-short-reads) - The peaks called ATAC reads (R2 reads).
* [short_reads](#short-reads) - The short reads.
* [long_reads](#long-reads) - The long interactions.
* [MAPSpeak](#tads-and-loops) - The TADs and loops determined by MAPS.
* [DEpeaks](#differential-analysis) Differential analysis for the TADs and loops.

## FastQC

[FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination and overrepresented sequences.

For further reading and documentation see the [FastQC help pages](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/).

**Output files:**

* `fastqc/`
  * `*_fastqc.html`: FastQC report containing quality metrics for your untrimmed raw fastq files.
* `fastqc/zips/`
  * `*_fastqc.zip`: Zip archive containing the FastQC report, tab-delimited data file and plot images.

> **NB:** The FastQC plots displayed in the MultiQC report shows _untrimmed_ reads. They may contain adapter sequence and potentially regions with low quality.

## MultiQC

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarizing all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability.

For more information about how to use MultiQC reports, see [https://multiqc.info](https://multiqc.info).

**Output files:**

* `multiqc/`
  * `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
  * `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
  * `multiqc_plots/`: directory containing static images from the report in various formats.

## Pipeline information

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.

**Output files:**

* `pipeline_info/`
  * Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  * Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.csv`.
  * Documentation for interpretation of results in HTML format: `results_description.html`.

## Genome

**Output files:**

* `genome/`
    * `genome.fa.sizes`: chromosome sizes file
    * `genome.fa.fai`: genome index file
    * `filtered/genome.include_regions.bed`: filtered genome by blacklist
    * `digest/*`: genomic features files digested by given restriction enzyme

If the mappability file is not provided in the profile file, the [GenMap](https://pubmed.ncbi.nlm.nih.gov/32246826/) will be used to create the mappability file with read length set to 50.

## Adapter trimming

[Cutadapt](http://dx.doi.org/10.14806/ej.17.1.200) removes adaptere sequences from high-throughput sequencing reads. The adapter pattern will be determined by the cutting sites of restriction sites.

**Output files:**

* `trimmed_fq/*`: trimmed fastq files

## Alignment

**Output files:**

* `bwa/mapped/*.bam`: The files resulting from the alignment of individual libraries will contain the coordinate sorted alignment files in [`*.bam`](https://samtools.github.io/hts-specs/SAMv1.pdf) format.

Adapter-trimmed reads are mapped to the reference assembly using [BWA::mem](http://bio-bwa.sourceforge.net/bwa.shtml). A genome index is required to run BWA so if this is not provided explicitly using the `--bwa_index` parameter then it will be created automatically from the genome fasta input. The index creation process can take a while for larger genomes.

## Pairs

The raw and filtered pairs exported by [pairtools](https://pairtools.readthedocs.io/en/latest/).
The quality analysis for filtered pairs were done by [pairsqc](https://github.com/4dn-dcic/pairsqc).

**Output files:**

* `pairs/`
    * `raw/*`: The raw reads pairs for each sample and stats for the pairs.
    * `filtered/*`: The filtered reads pairs (remove duplicated reads and validate pairs) for each sample and their index.
    * `QC/*`: The quality analysis results.

## Cooler

Cooler files for visualization.

**Output files:**

* `pairs/mcool/*`: The mcool files for each group.

## Call peaks for short reads

The peaks called ATAC reads (R2 reads) by MACS2.

* `ATACpeak/`
    * `shifted_R2_reads/*`: The shifted R2 reads for each sample.
    * `merged_R2_reads/*`: The shifted R2 reads for each group.
    * `peaks_per_Group/*`: The called peaks for each group.
    * `merged_peaks/*`: The merged peaks for all groups.

## Short reads

Short reads are extract from the shifted R2 reads of the pairs. The files are prepared for MAPS.

**Output files:**

* `short_reads/`
    * `byGroup/*`: The merged R2 reads for each group.
    * `bySample/*`: The R2 reads for each sample.

## Long reads

The reads pairs dumped by [Cooler](https://cooler.readthedocs.io/en/latest/index.html). The files are prepared for MAPS.

**Output files:**

* `long_reads/`
    * `byGroup/*`: The long distance reads pairs for each group.
    * `bySample/*`: The long distance reads pairs for each sample.

## TADs and loops

The TADs and loops are called by [MAPS](https://pubmed.ncbi.nlm.nih.gov/30986246/).
MAPS is a model-based analysis of long-range chromatin interactions.

**Output files:**

* `MAPSpeak/`
    * `parameters/*`: The parameter setting for MAPS.
    * `*`: The tables of TADs and loops with p-value and fdr.

## Differential analysis

Differential analysis for the TADs and loops by [baySeq](https://pubmed.ncbi.nlm.nih.gov/20698981/).
The input counts are filtered counts dumped by pairtools for each samples.
The differential analysis is done for validated TADs and loops by MAPs.

**Output files:**

* `DifferentialAnalyis/*`: Differential analysis results.
